lib/async_hooks.js

```diff
@@ -259,8 +262,15 @@ function validateAsyncId(asyncId, type) {
 
 // Embedder API //
 
+const undestroyedAsyncIds = new Set();
+function emitDestroyHook(asyncId) {
+  if (undestroyedAsyncIds.delete(asyncId)) {
+    emitDestroyScript(asyncId);
+  }
+}
+
 class AsyncResource {
-  constructor(type, triggerAsyncId = initTriggerId()) {
+  constructor(type, triggerAsyncId = initTriggerId(), opts = {}) {
```

Comment:
Given that that `triggerAsyncId` is also optional. I think the signature should be:

```js
new AsyncResource (type, options = {})
```

For backward compatibility, we can check with `typeof options === 'number'` and also deprecate that usage.